examples/go/tagging/gotag-fn/handler.go

```diff
@@ -0,0 +1,177 @@
+	Password string
+	Server   string
+	User     string
+}
+
+// Config represents the toml vcconfig file
+type vcConfig struct {
+	Tag     tagInfo
+	VCenter vCenterAuth
+}
+
+// The request from vmware-event-router.
+type dataObj struct {
+	Data nest1 `json:"data"`
+}
+type nest1 struct {
+	Vm nest2 `json:"Vm"`
+}
+type nest2 struct {
+	Vm vmInfo `json:"Vm"`
+}
+type vmInfo struct {
+	Value string `json:"Value"`
+	Type  string `json:"Type"`
+}
+
+// Handle a function invocation
+func Handle(req handler.Request) (handler.Response, error) {
+	ctx := context.Background()
+
+	// Retrieve vCenter and tag info from the config file.
+	vcc := loadTomlCfg()
+
+	// Incoming request object.
+	var mor dataObj
+	err := json.Unmarshal(req.Body, &mor)
+	if err != nil {
+		return handler.Response{
+			Body:       nil,
+			StatusCode: http.StatusBadRequest,
+		}, err
+	}
+
+	if mor.Data.Vm.Vm.Value == "" || mor.Data.Vm.Vm.Type == "" {
+		return handler.Response{
+			Body:       nil,
+			StatusCode: http.StatusBadRequest,
+		}, errors.New("managed Reference Object information must not be empty")
+	}
+
+	// Fill information in the request into a govmomi type.
+	var moRef types.ManagedObjectReference
+	moRef.Value = mor.Data.Vm.Vm.Value
+	moRef.Type = mor.Data.Vm.Vm.Type
+
+	// Connect to vcenter with govmomi client
+	log.Println("Connecting to vCenter")
+	clients := connect(ctx, vcc.VCenter)
+	log.Println("Connection successful.")
+
+	// tag the VM moRef with tagID (URN)
+	clients.tagVM(ctx, moRef, vcc.Tag.Urn)
+
+	// Respond with a message that tagging of vm was complete
+	message := fmt.Sprintf(
+		"%v was tagged with %v",
+		moRef.Value,
+		vcc.Tag.Urn,
+	)
+	log.Println(string(message))
+
+	err = clients.Govmomi.Logout(ctx)
+	if err != nil {
+		log.Printf("Unable to log out of govmomi due to %v.", err)
+	}
+	err = clients.Rest.Logout(ctx)
+	if err != nil {
+		log.Printf("Unable to log out of rest due to %v.", err)
+	}
+	log.Println("Logged out of Govmomi and Rest clients.")
+
+	return handler.Response{
+		Body:       []byte(message),
+		StatusCode: http.StatusOK,
+	}, nil
+}
+
+func loadTomlCfg() *vcConfig {
+	secretPath := "/var/openfaas/secrets/vcconfig"
+	secret, err := toml.LoadFile(secretPath)
+	if err != nil {
+		log.Println(err)
+	}
+
+	cfg := vcConfig{}
+	secret.Unmarshal(&cfg)
+
+	return &cfg
+}
+
+// Clients needed to access vSphere APIs
+type Clients struct {
+	Govmomi *govmomi.Client
+	Rest    *rest.Client
+}
+
+// Connect saves connection information to clients.
+func connect(ctx context.Context, vca vCenterAuth) *Clients {
+	var c Clients
+
+	u := url.URL{
+		Scheme: "https",
+		Host:   vca.Server,
+		Path:   "sdk",
+	}
+
+	log.Println("Get govmomi client.")
+	govc, err := govmomi.NewClient(ctx, &u, true)
+	if err != nil {
+		log.Println(err)
+	}
+	c.Govmomi = govc
+
+	log.Println("Login to govmomi client")
+	u.User = url.UserPassword(vca.User, vca.Password)
+
+	err = c.Govmomi.Login(ctx, u.User)
+	if err != nil {
+		log.Printf("ERROR: %v", err)
```

Comment:
inconsistent error printing. caps are usually not used in go. also consider wrapping and returning the error instead of handling them in these functions.

Reply:
Now that we're using Go 1.13, we can take advantage of the error package and the error.Wrap() function. I'll try to apply it everywhere I return an error--how exciting! 😸